src/internal/operators/bufferWhen.ts

```diff
@@ -48,89 +47,53 @@ import { SimpleOuterSubscriber, innerSubscribe, SimpleInnerSubscriber } from '..
  * @return {Observable<T[]>} An observable of arrays of buffered values.
  * @name bufferWhen
  */
-export function bufferWhen<T>(closingSelector: () => Observable<any>): OperatorFunction<T, T[]> {
-  return function (source: Observable<T>) {
-    return lift(source, new BufferWhenOperator(closingSelector));
-  };
-}
-
-class BufferWhenOperator<T> implements Operator<T, T[]> {
-  constructor(private closingSelector: () => Observable<any>) {}
-
-  call(subscriber: Subscriber<T[]>, source: any): any {
-    return source.subscribe(new BufferWhenSubscriber(subscriber, this.closingSelector));
-  }
-}
-
-class BufferWhenSubscriber<T> extends SimpleOuterSubscriber<T, any> {
-  private buffer: T[] | undefined;
-  private subscribing: boolean = false;
-  private closingSubscription: Subscription | undefined;
-
-  constructor(destination: Subscriber<T[]>, private closingSelector: () => Observable<any>) {
-    super(destination);
-    this.openBuffer();
-  }
-
-  protected _next(value: T) {
-    this.buffer!.push(value);
-  }
-
-  protected _complete() {
-    const buffer = this.buffer;
-    if (buffer) {
-      this.destination.next(buffer);
-    }
-    super._complete();
-  }
-
-  unsubscribe() {
-    if (!this.closed) {
-      this.buffer = null!;
-      this.subscribing = false;
-      super.unsubscribe();
-    }
-  }
-
-  notifyNext(): void {
-    this.openBuffer();
-  }
+export function bufferWhen<T>(closingSelector: () => ObservableInput<any>): OperatorFunction<T, T[]> {
+  return (source: Observable<T>) =>
+    lift(source, function (this: Subscriber<T[]>, source: Observable<T>) {
+      const subscriber = this;
+      let buffer: T[] | null = null;
+      let closingSubscriber: Subscriber<T> | null = null;
+      let isComplete = false;
 
-  notifyComplete(): void {
-    if (this.subscribing) {
-      this.complete();
-    } else {
-      this.openBuffer();
-    }
-  }
+      const openBuffer = () => {
+        closingSubscriber?.unsubscribe();
 
-  openBuffer() {
-    let { closingSubscription } = this;
+        const b = buffer;
+        buffer = [];
+        b && subscriber.next(b);
 
-    if (closingSubscription) {
-      this.remove(closingSubscription);
-      closingSubscription.unsubscribe();
-    }
+        let closingNotifier: Observable<any>;
+        try {
+          closingNotifier = from(closingSelector());
+        } catch (err) {
+          subscriber.error(err);
+          return;
+        }
 
-    const buffer = this.buffer;
-    if (this.buffer) {
-      this.destination.next(buffer);
-    }
+        closingNotifier.subscribe(
+          (closingSubscriber = new OperatorSubscriber(subscriber, openBuffer, undefined, () => {
+            isComplete ? subscriber.complete() : openBuffer();
```

Comment:
issue: The `subscriber.complete` call here seems unnecessary. `isComplete` is only set to `true` in the other `OperatorSubscriber` and immediately after that `subscriber.complete` is called. If there's a reason for this, maybe it warrants a comment?

Reply:
Good find.